### PR TITLE
IAM storage

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: src/storage/storage.cpp [hunk 1]
```diff
@@ -165,6 +170,81 @@ aos::RetWithError<aos::sm::servicemanager::ServiceData> Storage::GetService(cons
     return aos::RetWithError<aos::sm::servicemanager::ServiceData>(ConvertServiceData(serviceData));
 }
 
+aos::Error Storage::AddCertInfo(const aos::String& certType, const aos::iam::certhandler::CertInfo& certInfo)
+{
+    aos::LockGuard lock(mMutex);
+
+    LOG_DBG() << "Add cert info: " << certType;
+
+    return mCertDatabase.Add(ConvertCertInfo(certType, certInfo));
```
Comment (al1img) [inline src/storage/storage.cpp:179]: It looks like task added by  "[storage] Return error if record exist" commit, isn't properly implemented: we should return error if key field/fields already exists. Key field/fields are different in each DB. We need to reimplement it. I've reopened the task.

File: src/storage/storage.hpp [hunk 2]
```diff
@@ -178,13 +227,41 @@ class Storage : public aos::sm::launcher::StorageItf,
         }
     };
 
+    struct CertInfo {
+        uint8_t  mIssuer[aos::crypto::cCertIssuerSize];
+        size_t   mIssuerLen;
```
Comment (al1img) [inline src/storage/storage.hpp:232]: mIssuerSize
mSerialSize

We use size for bytes and len for string.

File: src/storage/storage.hpp [hunk 3]
```diff
@@ -178,13 +227,41 @@ class Storage : public aos::sm::launcher::StorageItf,
...
+
+        /**
+         * Compares cert info.
+         *
+         * @param cert info to compare.
+         * @return bool.
+         */
+        bool operator==(const CertInfo& rhs) const
+        {
+            return memcmp(mIssuer, rhs.mIssuer, sizeof(mIssuer)) == 0
```
Comment (al1img) [inline src/storage/storage.hpp:248]: first you should compare sizes and then do memcmp for size only.

File: src/storage/storage.hpp (hunk 2 above)
Comment (MykolaSolyanko) [reply src/storage/storage.hpp]: Done

File: src/storage/storage.cpp (hunk 1 above)
Comment (MykolaSolyanko) [reply src/storage/storage.cpp:179]: Ok

File: src/storage/storage.hpp (hunk 3 above)
Comment (MykolaSolyanko) [reply src/storage/storage.hpp:248]: Done

File: src/storage/storage.hpp [hunk 4]
```diff
@@ -178,13 +227,41 @@ class Storage : public aos::sm::launcher::StorageItf,
...
     };
 
+    struct CertInfo {
+        uint8_t  mIssuer[aos::crypto::cCertIssuerSize];
+        size_t   mIssuerLen;
+        uint8_t  mSerial[aos::crypto::cCertSerialNumberSize];
+        size_t   mSerialLen;
+        char     mCertURL[aos::cURLLen];
+        char     mKeyURL[aos::cURLLen];
+        char     mCertType[aos::iam::certhandler::cCertTypeLen];
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/storage/storage.hpp]: Done

**Review by al1img — APPROVED** [review]

**Pull request merged**